### ISSUE TITLE
ASL loss=0 during training

### ISSUE THREAD

**Issue by chijames** [harsh19/SPINE#2]
I run the code using the recommended command/setting on the readme page. However, the trained embeddings are not as good as the provided ones when visualizing them using the provided script(both glove/word2vec), and I also notice that the ASL loss is always zero throughout the training process. Do you have any suggestion? Thanks!

**Comment by danishpruthi** [comment]
Thanks for trying out, and pointing this out - we really appreciate it. Actually, a few small errors creeped in when we were porting our code to pytorch for a public release, we are sorry for the inconvenience. I have [fixed the problem](https://github.com/harsh19/SPINE/commit/cc33f5029141ab2aa6c88cd2f60f02fbbb13eeee) in the ASL, and we erroneously recommended using 500 epochs, while consistently 4000 epochs are better (and required). Further, for word2vec we find 0.2 noise level as best, whereas glove requires noise level of 0.4 

I ran the code again, and now the representations are very similar to the provided ones.

**Comment by chijames** [comment]
Thanks for the quick reply! After running the updated code with the new recommended setting, the ASL loss is non-negative for the glove embedding. However, the ASL loss for word2vec embedding still seems to be zero and the final sparsity is about 99.9, which I believe is much larger than the desired value 0.85. Just wondering if this is normal? Thanks.

**Comment by danishpruthi** [comment]
ASL, as per definition, doesn't penalize the model for producing representations sparser than the desired number. If your use case requires it to be otherwise, you can easily tweak the loss to not do that. If your reconstruction loss is low enough (say < 0.01, which means that on average the deviation from the input is only 0.1 per value in a word vector): then extra sparsity helps. It is less desirable if the additional sparsity comes at the cost of reconstruction loss (in which case, you can weight the reconstruction loss coefficient higher than the ASL and PSL loss).
